base/util/PKICertImport.bash

```diff
@@ -0,0 +1,333 @@
+        local ret=0
+        local verify_args=("-V")
+
+        verify_args+=("-d" "$NSSDB_TYPE$database")
+        verify_args+=("-n" "$CERT_NICKNAME")
+        verify_args+=("-u" "$CERT_USAGE")
+        if [ "x$HSM_TOKEN" != "x" ]; then
+            verify_args+=("-h" "$HSM_TOKEN")
+        fi
+        if [ "x$NSSDB_PASSWORD" != "x" ]; then
+            verify_args+=("-f" "$NSSDB_PASSWORD")
+        fi
+
+        # Ensures that the signature is checked as well.
+        verify_args+=("-e")
+
+        # Validate the certificate. Note that we have to pattern match on the
+        # output of certutil -V; the return code is uncorrelated with the
+        # actual result. (It is dependent upon whether or not a HSM is used).
+        __v certutil "${verify_args[@]}"
+
+        local certutil_result=""
+        certutil_result="$(certutil "${verify_args[@]}" 2>&1)"
+
+        grep -q '^certutil: certificate is valid$' <<< "$certutil_result"
+        ret=$?
+
+        if (( ret != 0 )); then
+            __e "$certutil_result" 1>&2
+        fi
+
+        return $ret
+    }
+
+    # Remove the certificate from the NSS DB specified by $1. Errors are fatal;
+    # uses exit instead of return.
+    function _remove_cert() {
+        local database="$1"
+        local remove_args=("-D")
+
+        remove_args+=("-d" "$NSSDB_TYPE$database")
+        if [ "x$NSSDB_PASSWORD" != "x" ]; then
+            remove_args+=("-f" "$NSSDB_PASSWORD")
+        fi
+
+        __v certutil "${remove_args[@]}" "-n" "$CERT_NICKNAME"
+        certutil "${remove_args[@]}" "-n" "$CERT_NICKNAME"
+        local ret=$?
+        if (( ret != 0 )); then
+            __e "certutil returned non-zero result: $ret"
+            __e "Unable to delete certificate!"
+            exit $?
+        fi
+
+        if [ "x$HSM_TOKEN" != "x" ]; then
+            # In the event we have a HSM, we also have to remove it from the
+            # HSM token.
+
+            __v certutil "${remove_args[@]}" "-n" "$HSM_TOKEN:$CERT_NICKNAME"
+            certutil "${remove_args[@]}" "-n" "$HSM_TOKEN:$CERT_NICKNAME"
+            local ret=$?
+            if (( ret != 0 )); then
+                __e "certutil returned non-zero result: $ret"
+                __e "Unable to delete certificate!"
+                exit $?
+            fi
+        fi
+    }
+
+    ## [ program flow ] ##
+    local ret=0
+
+    # The program flow of this script is:
+    #
+    # - Parse arguments
+    #   - [print help if required]
+    # - Create working directory
+    # - Copy files into the working directory
+    # - Import certificate into NSS DB
+    # - Validate certificate in NSS DB
+    #   - [on failure, remove from NSS DB]
+    #
+    # Import is handled by _import_cert, verification is handled by
+    # _verify_cert, and removal is handled by _remove_cert.
+
+    _parse_args "$@"
+    ret="$?"
+    if (( ret != 0 )); then
+        _print_help $ret
+        exit 1
+    fi
+
+    _import_cert "$NSSDB"
```

Comment:
@ladycfu At the expense of making "import to a second soft token" harder if we do need it in the future, we can remove the arguments to these and just use the global `$NSSDB` variable inside the functions. Thoughts?

Reply:
I"m okay if you just leave them in.